src/lib/operators/abstract_operator.cpp

```diff
@@ -122,19 +163,53 @@ void AbstractOperator::execute() {
 
 std::shared_ptr<const Table> AbstractOperator::get_output() const { return _output; }
 
-void AbstractOperator::clear_output() { _output = nullptr; }
+void AbstractOperator::clear_output() {
+  Assert(_executed, "Unexpected call of clear_output() since operator did not execute yet.");
+  Assert(_consumer_count == 0, "Cannot clear output since there are still consuming operators.");
+  if (_clear_output) _output = nullptr;
+}
 
 std::string AbstractOperator::description(DescriptionMode description_mode) const { return name(); }
 
 std::shared_ptr<AbstractOperator> AbstractOperator::deep_copy() const {
   std::unordered_map<const AbstractOperator*, std::shared_ptr<AbstractOperator>> copied_ops;
-  return _deep_copy_impl(copied_ops);
+  return deep_copy(copied_ops);
+}
+
+std::shared_ptr<AbstractOperator> AbstractOperator::deep_copy(
+    std::unordered_map<const AbstractOperator*, std::shared_ptr<AbstractOperator>>& copied_ops) const {
+  const auto copied_ops_iter = copied_ops.find(this);
+  if (copied_ops_iter != copied_ops.end()) return copied_ops_iter->second;
+
+  const auto copied_left_input =
+      left_input() ? left_input()->deep_copy(copied_ops) : std::shared_ptr<AbstractOperator>{};
+  const auto copied_right_input =
+      right_input() ? right_input()->deep_copy(copied_ops) : std::shared_ptr<AbstractOperator>{};
+
+  auto copied_op = _on_deep_copy(copied_left_input, copied_right_input, copied_ops);
+  if (_transaction_context) copied_op->set_transaction_context(*_transaction_context);
```

Comment:
Not sure. I did not change anything. I just moved the code from the bottom to the top.

Reply:
Thought about it some more: We rely on this, e.g., in the ExpressionEvaluator when we duplicate a correlated PQP and set the parameters. There, we do not explicitly set the tx context. Maybe add a comment?

> Set the transaction context so that we can execute the copied plan in the current transaction (see, e.g., ExpressionEvaluator::_evaluate_subquery_expression_for_row)

Reply:
Added.